### PR TITLE
wip: explicit spans for async

### PR DESCRIPTION
**Pull request by c-cube**
- doc
- gitignore
- Add explicit spans, for async tracing.
- update test

### CODE REVIEW

**Comment by c-cube** [comment]
cc @ELLIOTTCABLE

**Review by ELLIOTTCABLE — CHANGES_REQUESTED** [review]
Will try and review the rest of this later, but noticed this — intentional?

**Comment by c-cube** [comment]
Yes, intentional. For a user facing API, optional args can be useful,
but for the collector I'd rather be as explicit as possible these days.

File: src/core/trace_core.mli [hunk 1]
```diff
@@ -47,7 +25,33 @@ val with_span :
...
+
+val enter_explicit_span :
+  surrounding:explicit_span option ->
+  ?__FUNCTION__:string ->
+  __FILE__:string ->
+  __LINE__:int ->
+  ?data:(unit -> (string * user_data) list) ->
+  string ->
+  explicit_span
+(** Like {!enter_span} but the caller is responsible for
```
Comment (tatchi) [inline src/core/trace_core.mli:45]:
```suggestion
(** Like {!with_span} but the caller is responsible for
```

File: src/core/trace_core.mli [hunk 2]
```diff
@@ -11,43 +12,60 @@ val enabled : unit -> bool
...
+    This is the recommended way to instrument most code.
 
-    This should be called exactly once per span (even in case of exception).
-    Once this is called, a timestamp might be recorded,
-  and the span is considered finished and can't be used anymore *)
+    {b NOTE} an important restriction is that this is only supposed to
+    work for synchronous, direct style code. Monadic concurrency, Effect-based
+    fibers, etc. might not play well with this style of spans on some
+    or all backends. If you use cooperative concurrency,
+    see {!enter_manual_span}.
```
Comment (ELLIOTTCABLE) [inline src/core/trace_core.mli:34]: I'd definitely rewrite this; but it's a subtle point, and I'm not sure _exactly_ what to say instead.

Now that the API has been redesigned, `enter_explicit_span` and `with_span` should both work perfectly fine in both direct-style/on-the-stack _and_ promise/async-style/event-loop code just fine! The old 'uh-oh, implicit parent-tracking is entirely borked if you're in an event-loop' is _fixed_ now, specifically _because_ we add these functions with non-optional `parent` demands surfaced to the user. They "know" (in the sense of 'are forced to deal with') the shortcomings of the storage options if they choose functions that take explicit arguments, that they then must expose to their users. … right? :x

Really the only major caveat at this point, I think? unless I've become bogged down? is that `enter_initial_span` should prolly not be used by "libraries", only by top-level-apps, if you  want to be a good citizen.

File: src/core/trace_core.mli [hunk 3]
```diff
@@ -11,43 +12,60 @@ val enabled : unit -> bool
...
   string ->
-  (span -> 'a) ->
-  'a
-(** [with_span ~__FILE__ ~__LINE__ name f] enters a new span [sp],
-    and calls [f sp].
-    [sp] might be a dummy span if no collector is installed.
-    When [f sp] returns or raises, the span [sp] is exited.
+  explicit_span
+(** Like {!with_span} but the caller is responsible for
+    providing the [parent] context, and carry the resulting
```
Comment (ELLIOTTCABLE) [inline src/core/trace_core.mli:46]: ditto — i'd make the 'by exposing it in your library API' part explicit. give library-authors explicit directions, so they don't have to think:

```suggestion
    obtaining the [parent] span from their {e own} caller, and carry the resulting
```

File: src/core/trace_core.mli [hunk 4]
```diff
@@ -11,43 +12,60 @@ val enabled : unit -> bool
...
+  ?__FUNCTION__:string ->
+  __FILE__:string ->
+  __LINE__:int ->
+  ?data:(unit -> (string * user_data) list) ->
+  string ->
+  explicit_span
+(** Like {!with_span} but the caller is responsible for carrying this
+    [explicit_span] around until it's exited with {!exit_manual_span}.
+    The span can be used as a parent in {!enter_manual_sub_span}.
+    @since NEXT_RELEASE *)
```
Comment (ELLIOTTCABLE) [inline src/core/trace_core.mli:60]: said it above, but in the wrong place, lmao — i suggest discouraging use of this. generally speaking, it's pretty catastrophic in UIs if root-level spans start appearing from deep dependencies you don't control … which is, to be clear, a shortcoming of the UIs I've used, but unfortunately still a real-world thing, haha.

tl;dr sth like "this should probably only be used if you're instrumenting a top-level application with Trace; library authors should stick to `with_span` to obtain implicit parentage from their dependants, or `enter_manual_sub_span`, with a `parent` obtained from their dependant/caller."

File: src/core/types.ml [hunk 5]
```diff
@@ -11,3 +11,11 @@ type user_data =
   ]
 (** User defined data, generally passed as key/value pairs to
     whatever collector is installed (if any). *)
+
+type explicit_span = {
+  span: span;
+      (** Identifier for this span. Several explicit spans might share the same
+      identifier since we can differentiate between them via [meta]. *)
+  meta: Meta_map.t;  (** Metadata for this span (and its context) *)
+}
```
Comment (ELLIOTTCABLE) [inline src/core/types.ml:20]: do we want this to be public/constructable? perhaps it should be opaque to both collectors and library-authors, so that we can control the impl?

File: test/t1.ml [hunk 6]
```diff
@@ -6,12 +6,28 @@ let run () =
...
+        (* fake micro sleep *)
+        let _sp =
+          Trace.enter_manual_sub_span ~parent:pseudo_async_sp ~__FILE__
+            ~__LINE__ "sub-sleep"
+        in
+        Thread.delay 0.005;
+        Trace.exit_manual_span _sp
+      ) else if _j = 3 then
+        (* pretend some task finished. Note that this is not well scoped wrt other spans. *)
+        Trace.exit_manual_span pseudo_async_sp
```
Comment (ELLIOTTCABLE) [inline test/t1.ml:30]: 👏 I think this turned out really well, tbh

File: src/core/collector.ml [hunk 7]
```diff
@@ -9,33 +9,63 @@ open Types
...
+    'a
+  (** Run the function in a new span.
+
+      This replaces the previous [enter_span] and [exit_span] which were too flexible
+      to be efficient to implement in async contexts.
+     @since NEXT_RELEASE *)
+
+  val enter_manual_span :
+    parent:explicit_span option ->
     ?__FUNCTION__:string ->
```
Comment (ELLIOTTCABLE) [inline src/core/collector.ml:36]: shouldn't this, then, _not_ be optional now, to match?

```suggestion
    __FUNCTION__:string option ->
```

**Review by ELLIOTTCABLE — CHANGES_REQUESTED** [review]
mostly doc comments, although you missed a `?__FUNCTION__`.

yesssssss

File: test/t1.ml (hunk 6 above)
Comment (c-cube) [reply test/t1.ml:30]: :)

File: src/core/types.ml (hunk 5 above)
Comment (c-cube) [reply src/core/types.ml:20]: At least in `trace-tef` it's already used to get/set data in `meta` (to transmit that data from the enter to the exit, basically). We need to at least expose get/set meta functions if we make this private.

On the face of it I do like the idea of making this private though.

Comment (ELLIOTTCABLE) [reply src/core/types.ml:20]: Would it be possible to lock it down entirely, and expose _specifically_ what `tef` needs; then leave it as a 'talk to us if you need us to expose more `meta` interfaces to collectors' kinda thing?

I'm just concerned about providing too much, too early, and then being locked out of future changes — backward- and forward-compatibility are _extremely_ important for this project, for obvious reasons.

Comment (c-cube) [reply src/core/types.ml:20]: I'm worried that, a few years down the line, we have dozens (hundreds?) of bits of metadata exposed here because `trace` becomes a chokepoint for all communications between the user and the collector.
